fibers/operations.scm

```diff
@@ -67,14 +67,20 @@
 (define-inlinable (make-op-state) (make-atomic-box 'W))
 
 (define-record-type <base-op>
-  (make-base-operation wrap-fn try-fn block-fn)
+  (%make-base-operation wrap-fn try-fn block-fn cancel-fn)
   base-op?
   ;; ((arg ...) -> (result ...)) | #f
   (wrap-fn base-op-wrap-fn)
   ;; () -> (thunk | #f)
   (try-fn base-op-try-fn)
   ;; (op-state sched resume-k) -> ()
-  (block-fn base-op-block-fn))
+  (block-fn base-op-block-fn)
+  ;; (sched) -> ()
+  (cancel-fn base-op-cancel-fn))
+
+(define* (make-base-operation wrap-fn try-fn block-fn
+                              #:optional (cancel-fn (const #f)))
```

Comment:
Going by other comments, I think it's too soon to tell whether this will be _the_ API for registering clean-up functions.
So, we should prepare for a potential API break, and make it a _clean_ API break. E.g. if the optional argument is replaced by some other argument with different semantics, that's potentially going to lead to unclear errors (or worse, undetected errors). This could be avoided by making it a keyword argument instead, or a `make-base-operation/with-cancellation` variant.

Reply:
Agreed.  In v2, there's a new `make-base-operation/internal` and `make-base-operation` is unchanged.